tokenserver-auth/src/lib.rs

```diff
@@ -75,36 +57,37 @@ impl Tokenlib {
         plaintext: MakeTokenPlaintext,
         shared_secret: &str,
     ) -> Result<(String, String), TokenserverError> {
-        Python::with_gil(|py| {
-            // `import tokenlib`
-            let module = PyModule::import(py, "tokenlib").map_err(|e| {
-                e.print_and_set_sys_last_vars(py);
-                e
-            })?;
-            // `kwargs = { 'secret': shared_secret }`
-            let kwargs = [("secret", shared_secret)].into_py_dict(py);
-            // `token = tokenlib.make_token(plaintext, **kwargs)`
-            let token = module
-                .getattr("make_token")?
-                .call((plaintext,), Some(kwargs))
-                .map_err(|e| {
-                    e.print_and_set_sys_last_vars(py);
-                    e
-                })
-                .and_then(|x| x.extract())?;
-            // `derived_secret = tokenlib.get_derived_secret(token, **kwargs)`
-            let derived_secret = module
-                .getattr("get_derived_secret")?
-                .call((&token,), Some(kwargs))
-                .map_err(|e| {
-                    e.print_and_set_sys_last_vars(py);
-                    e
-                })
-                .and_then(|x| x.extract())?;
-            // `return (token, derived_secret)`
-            Ok((token, derived_secret))
+        #[derive(Serialize)]
+        struct Token<'a> {
+            #[serde(flatten)]
+            plaintext: MakeTokenPlaintext,
+            salt: &'a str,
+        }
+
+        let mut salt_bytes = [0u8; 3];
+        let mut rng = rand::thread_rng();
+        rng.fill_bytes(&mut salt_bytes);
+        let salt = hex::encode(salt_bytes);
+        let token_str = serde_json::to_string(&Token {
+            plaintext,
+            salt: &salt,
         })
-        .map_err(pyerr_to_tokenserver_error)
+        .map_err(|_| TokenserverError::internal_error())?;
+        let crypto_lib = CryptoImpl {};
+        let hmac_key = crypto_lib.hkdf(shared_secret, None, HKDF_SIGNING_INFO)?;
+        let signature = crypto_lib.hmac_sign(&hmac_key, token_str.as_bytes())?;
+        let mut token_bytes = Vec::with_capacity(token_str.len() + signature.len());
+        token_bytes.extend_from_slice(token_str.as_bytes());
+        token_bytes.extend_from_slice(&signature);
+        let token = base64::engine::general_purpose::URL_SAFE.encode(token_bytes);
+        // Now that we finialized the token, lets generate our per token secret
+        let mut info = Vec::with_capacity(HKDF_INFO_DERIVE.len() + token.as_bytes().len());
+        info.extend_from_slice(HKDF_INFO_DERIVE);
+        info.extend_from_slice(token.as_bytes());
+
+        let per_token_secret = crypto_lib.hkdf(shared_secret, Some(salt.as_bytes()), &info)?;
+        let per_token_secret = base64::engine::general_purpose::URL_SAFE.encode(per_token_secret);
+        Ok((token, per_token_secret))
```

Comment:
Ditto a careful review of this, this is a fairly mechanical translation from tokenlib, but is still calling into crypto bits

Reply:
Also note to self to add some tests with test vectors that verify that a token generated using this function, can have its signature verified by any party that also has the signing secret

Reply:
I'm not sure how feasible it might be, but one of the things we did with Autopush was take a value generated using the old python code and see if we can match that value using rust. (e.g. you could generate a signature using a set of known values using the existing python and validate the rust by producing a matching signature, using the same values as inputs.)

Reply:
That's a great suggestion, I think that's what the end-to-end tests were trying to do in https://github.com/mozilla-services/syncstorage-rs/pull/1513#discussion_r1456756218

The signatures didn't match at first, but I debugged it was because the python library's json serializer keeps extra spaces that `serde_json` doesn't 😬 Calling the python library's signature function does create matching signatures and the end-to-end tests pass